File: examples/SparkvsMahout/SparkCollaborative/ScalaCollaborative/src/main/scala/com/example/CollaborativeFiltering.scala
package com.example


import java.io.FileWriter

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.SparkContext._
import org.apache.spark.mllib.recommendation.{ALS, Rating}

object CollaborativeFiltering {

  def main(args: Array[String]): Unit = {
    // Load and parse the data
    val conf = new SparkConf().setAppName("Scala Collaborative").setMaster("local")
    val sc = new SparkContext(conf)

    val data = if(args.isEmpty) sc.textFile("data/ml-100k/ua.base3")  else sc.textFile(args(0))

    val ratings = data.map(_.split(',') match { case Array(user, item, rate) =>
      Rating(user.toInt, item.toInt, rate.toDouble)
    })

    // Build the recommendation model using ALS
    val rank = 10
    val numIterations = 20

    var time = System.currentTimeMillis;
    val model = ALS.train(ratings, rank, numIterations, 0.01)
    time =  System.currentTimeMillis - time

    val fw = new FileWriter("data/output.txt", false) ;

    fw.write("Execution time for training: " + time + "\n")
    fw.close()
  }
}

File: examples/SparkvsMahout/SparkCollaborative/JavaCollaborative/src/main/java/com/example/CollaborativeFiltering.java
package com.example;

/**
 * Created by viana on 04/03/15.
 */
import scala.Tuple2;

import org.apache.spark.api.java.*;
import org.apache.spark.api.java.function.Function;
import org.apache.spark.mllib.recommendation.ALS;
import org.apache.spark.mllib.recommendation.MatrixFactorizationModel;
import org.apache.spark.mllib.recommendation.Rating;
import org.apache.spark.SparkConf;

import java.io.*;


public class CollaborativeFiltering {
    public static void main(String[] args) {
        SparkConf conf = new SparkConf().setAppName("Collaborative Filtering Example").setMaster("local");
        JavaSparkContext sc = new JavaSparkContext(conf);

        String path = (args.length > 0 && !args[0].isEmpty()) ? args[0] : "data/ua.base";
        String outputPath = (args.length > 1 && !args[1].isEmpty()) ? args[1] : "data/output";

        JavaRDD<String> data = sc.textFile(path);
        JavaRDD<Rating> ratings = data.map(
                new Function<String, Rating>() {
                    public Rating call(String s) {
                        String[] sarray = s.split(",");
                        return new Rating(Integer.parseInt(sarray[0]), Integer.parseInt(sarray[1]),
                                Double.parseDouble(sarray[2]));
                    }
                }
        );

        // Build the recommendation model using ALS
        int rank = 10;
        int numIterations = 20;
        double alpha = 0.01;
        long startTime = System.currentTimeMillis();

        MatrixFactorizationModel model = ALS.train(JavaRDD.toRDD(ratings), rank, numIterations, alpha);

        long endTime = System.currentTimeMillis();

        PrintWriter writer;
        try {
		System.out.println("=================================Time for model training in Java Spark: "+ (endTime - startTime) + " milli seconds");
            writer = new PrintWriter(new FileOutputStream(new File(outputPath), false));
            writer.println("Time for model training in Java Spark: "+ (endTime - startTime)
                    + " milli seconds");
            writer.close();
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
    }
}
